### PR TITLE
Improve presentation in Google sheets (with Braille mode enabled)

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
None.

### Summary of the issue:
Google has recently introduced a braille mode for Google Sheets, which exposes actual content via browser accessibility, rather than just forcing the AT to speak everything with a live region.
However, this new mode, uses a contenteditable table, thus NVDA tends to announce way too much inforamtion when navigating around. For instance, when moving down a row, NVDA announces the entire row, and also repetes the content of the cell several times. n Firefox it also announces x of y position info on the cell which is yet more pointless information.

### Description of how this pull request fixes the issue:
This PR ensures that if an IAccessible NvDAObject has the readonly state, any existing editable state is removed. In other words, if an IAccessible2 implementation exposes both editable and readonly states, readonly will win. This was a compromize with Google for them to be able to use contenteditable on the table, they also expose the readonly state to convey the editable state should be overridden.
Handling the readonly state in this way ensures that NVDA no longer reports a readonly stable cell in a contenteditable table as being editable text. I.e. it no longer tries to announce the line of text at the caret on the cell.
Also, table rows in IAccessible2 table implementations are no longer treeted as being presentable focus ancestors. This stops the reporting of the entire row when moving from one to another. The cells themselves have enough information.
The PR also adds support for IAccessible2's rowtext and coltext attributes, by exposing them via an NVDAObject's cellCoordsText if one or both of these attributes exist. They are exposed simply as a concatination of colText and rowText, similar to how cell coordinates are presented in MS Excel. Thus Google sheets cell coordinates now are presented the same as MS excel.
Finally,  IAccessible NvDAObjects no longer expose x of y position info if this object supports the IAccessibleTableCell interface. 2d table cell coordinates are much better than 1d index info.
 
### Testing performed:
Navigated around a spreadsheet in Google Sheets with braille mode enabled. Verified the correct information was exposed in both speech and braille.

### Known issues with pull request:
None.

### Change log entry:
New features:
* Early support for Google Sheets with Braille mode enabled.

### CODE REVIEW

**Comment by leonardder** [comment]
I wonder whether this also works for sheets that have been imported from Microsoft Excel. I belief they are exposed in a somewhat different editor.

File: source/NVDAObjects/IAccessible/__init__.py [hunk 1]
```diff
@@ -392,7 +392,7 @@ def findOverlayClasses(self,clsList):
 			clsList.insert(0, FocusableUnfocusableContainer)
 
 		if hasattr(self, "IAccessibleTextObject"):
-			if role==oleacc.ROLE_SYSTEM_TEXT or self.IA2States&IAccessibleHandler.IA2_STATE_EDITABLE:
+			if role==oleacc.ROLE_SYSTEM_TEXT or controlTypes.STATE_EDITABLE in self.states:
```
Comment (leonardder) [inline source/NVDAObjects/IAccessible/__init__.py]: I'm a bit bothered about this one. Are you sure that this won't cause side effects in totally unrelated applications?

**Comment by michaelDCurran** [comment]
That specific change just checks for the editable state in a more 
abstract way, so that  code has a chance to alter the states before the 
check.
self.states is made up of IAccessibleStates and IA2States. And now, also 
checks to ensure that readonly overrides editable.
Instead of checking self.states I could check both self.IA2States for 
editable and self.IAccessibleStates for not readonly... but that does 
seem like code duplication.
Can you elaborate on your concern where this could break?
I'm imagining you're worried that some subclass of IAccessible 
NvDAObject has an implementation of states that adds the editable state, 
yet  does not expect that the EditableText behaviour would be used.

**Comment by leonardder** [comment]
[**@michaelDCurran**](https://github.com/michaelDCurran) commented on [23 Jan 2018, 08:53 CET](https://github.com/nvaccess/nvda/pull/7935#issuecomment-359705514 "2018-01-23T07:53:46Z - Replied by Github Reply Comments"):
> I'm imagining you're worried that some subclass of IAccessible 
> NvDAObject has an implementation of states that adds the editable state, 
> yet  does not expect that the EditableText behaviour would be used.

Yes, that was exactly my concern. However, I missed that this behaviour will still be limited to IAccessible2 due to this logic only being applied when there is an IAccessibleTextObject, so I don't think it will cause anomalies.

**Comment by leonardder** [comment]
I've tried to test this on two machines. Though I've been able to enable braille support, it seems that either using Next or this branch from source does not make any difference in the behavior. Could it be that there are different versions of this feature in circulation?

**Comment by michaelDCurran** [comment]
It will be most noticeable in Firefox.

However, it is possible that this is still only available to 
participants in the Google Tester program. I thought perhaps it had 
already been released. But, as you say, you do see the option to enable 
braille mode... so in that case I'm not really sure what that does on 
older versions.

In a Google sheets spreadsheet, you should notice that NVDA will no 
longer announce the text of the entire row when moving up and down a 
column, and that it will not repete a lot of information on each cell. 
also, Cell coordinates are now things like a1, not row 1 column 1.

File: source/NVDAObjects/__init__.py [hunk 2]
```diff
@@ -643,15 +643,6 @@ def _get_presentationType(self):
 				return self.presType_layout
 			if role in (controlTypes.ROLE_TABLEROW,controlTypes.ROLE_TABLECOLUMN,controlTypes.ROLE_TABLECELL) and (not config.conf["documentFormatting"]["reportTables"] or not config.conf["documentFormatting"]["reportTableCellCoords"]):
 				return self.presType_layout
-		if role in (controlTypes.ROLE_TABLEROW,controlTypes.ROLE_TABLECOLUMN):
```
Comment (leonardder) [inline source/NVDAObjects/__init__.py]: May be you could elaborate on why you removed this? I see some of this logic has been moved to ia2Web, but couldn't there be cases outside ia2Web (i.e. UIA) in which case ROLE_TABLEROW and ROLE_TABLECOLUMN are mapped to objects, in which case these objects will be of presentation type content erroneously?

**Comment by leonardder** [comment]
[**@michaelDCurran**](https://github.com/michaelDCurran) commented on [23 jan. 2018 09:37 CET](https://github.com/nvaccess/nvda/pull/7935#issuecomment-359717149 "2018-01-23T08:37:40Z - Replied by Github Reply Comments"):
> In a Google sheets spreadsheet, you should notice that NVDA will no 
> longer announce the text of the entire row when moving up and down a 
> column, and that it will not repete a lot of information on each cell. 
> also, Cell coordinates are now things like a1, not row 1 column 1.

I'm not always getting whole row announcements, but keep getting selection changed messages. Furthermore, I'm getting row 1 column 1 announcements when arrowing through the table. Object navigation reveals proper coordinates text for every cell though.

**Review by feerrenrut — APPROVED** [review]

**Comment by leonardder** [comment]
Looks like braille mode has entirely disappeared from my google sheets.
[**@michaelDCurran**](https://github.com/michaelDCurran) commented on [23 jan. 2018 07:23 CET](https://github.com/nvaccess/nvda/pull/7935#issue-290722165 "2018-01-23T06:23:42Z - Replied by Github Reply Comments"):
> ### Change log entry:
> New features:
> * Improved support for Google Sheets with Braille mode enabled.

May be reword this to early support, similar to what has been done with Kindle math support in 2017.4?

**Comment by PratikP1** [comment]
According to an email I saw yesterday from the Docs PM on the Google accessibility discussion list, the team has temporarily disabled Braille support for Sheets for all screen readers other than Chromevox because of various issues that people were having with the interface.
